docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,233 @@
+
+#### Story 3
+
+As an application administrator, I want to make the **horizontal scaling** of my app more deterministic by submitting reservations before a scale-up. Besides, I can also reserve resources after a scale-down for future demands. It is useful especially when we want a guaranteed scale-up of applications for the coming business peak.
+
+#### Story 4
+
+As a cluster administrator, I want to **pre-allocate** node resources for future usage no matter whether they are available now or not. I want to allocate the future free resources but do not disrupt the running of scheduled pods. Reservation can be made to pre-allocate resources since it makes no physical cost to the node. It may be in a `Waiting` state. When there is enough space for the reservation, it will become `Available` for the owner pods' scheduling.
+
+### API
+
+In this section, a Custom Resource Definition (CRD) named `Reservation` is proposed to allow the scheduler to reserve node resources for specific pods.
+
+![image](/docs/images/resource-reservation.svg)
+
+```go
+type Reservation struct {
+	metav1.TypeMeta `json:",inline"`
+	// <unresolved>
+	// Should the reservation be namespaced?
+	// </unresolved>
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ReservationSpec   `json:"spec,omitempty"`
+	Status            ReservationStatus `json:"status,omitempty"`
+}
+
+type ReservationSpec struct {
+	// Template defines the scheduling requirements (resources, affinities, images, ...) processed by the scheduler just like a normal pod.
+	// If the `template.spec.nodeName` is specified, the scheduler will not choose another node but reserve resources on the specified node.
+	Template *corev1.PodTemplateSpec `json:"template,omitempty"`
+	// Select the owners that can allocate the reserved resources.
+	// Multiple selectors and ANDed.
+	OwnerSelectors []ReservationOwnerSelector `json:"ownerSelectors,omitempty"`
+	// Time-to-Live period (in seconds) for the reservation.
+	// If both `TTLSeconds` and `ExpiredTimestamp` are not specified, a very long TTL will be picked by default.
+	TTLSeconds *int64 `json:"ttlSeconds,omitempty"`
+	// Expired time for the reservation.
+	// If specifed, the `TTLSeconds` will be ignored.
+	ExpiredTimestamp metav1.Time `json:"expiredTimestamp,omitempty"`
+	// If specified, allocate resources from the referenced objects.
+	ResourceFrom []ReservationOwner `json:"resourceFrom,omitempty"`
+	// By default, reservation's resources requirements (specified in `template.spec`) sums resources would get eventually (specified in `ResourceFrom`) should be not larger than the free resources of the node.
+	// Once `SkipResourceValidate` is set, this validation will be skipped and then overcommitment is allowed.
+	SkipResourceValidate bool `json:"skipResourceValidate,omitempty"`
```

Comment:
When we specify the resourceFrom, is it equivalent to pre-allocation, and this field is not needed?

Reply:
Yes when the `resourceFrom` is specifed. And `SkipResourceValidate` works when we do not specify `resourceFrom` but expect a pre-allocation.

Reply:
We can simplify the pre-allocation procedure, and the `resourceFrom` field is not needed.  
When we want to pre-allocate resources from some nodes, we can set `PreAllocation` to true indicating that the resource requirements of the reservation are ignored in the Filter phase and the reservation is `Waiting` to be `Available` until there are enough free resources on the node. Additionally, we can set the `template` with affinity/anti-affinity to schedule reservations on specified nodes (maybe there are specified pods).